src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java

```diff
@@ -132,24 +134,26 @@
   /**
    * The requested server version.
    */
-  private ExchangeVersion requestedServerVersion = ExchangeVersion.Exchange2010_SP2;
+  private final ExchangeVersion requestedServerVersion;
 
   /**
    * The server info.
    */
   private ExchangeServerInfo serverInfo;
 
-  private Map<String, String> httpHeaders = new HashMap<String, String>();
-
-  private Map<String, String> httpResponseHeaders = new HashMap<String, String>();
+  private Map<String, String> httpHeaders = new ConcurrentHashMap<String, String>();
```

Comment:
The way ExchangeService is implemented today it isn't thread safe even when you create new instances of it for each thread - see #371.

If it is thread safe (as this PR makes it), ExchangeService becomes:
# Easier to use. I believe developers would assume it's thread safe, as *Service named classes usually are. It's also less code to use it thread safe rather than creating, configuration, then closing it for each use.
# Faster. Before this PR, the suggested proper use of ExchangeService is to create and then tears down an HttpClient every time it's used - that takes time. It's also unable to share HttpClient resources (like thread pools) as the HttpClient design intends, reducing performance.
# Less memory intensive. By having fewer instances of ExchangeService, less memory is used. Most significantly, by sharing HttpClient instances, thread pools, caches, and objects are all shared (as HttpClient is intended to be used).